opengrok-indexer/src/main/java/org/opengrok/indexer/web/PageConfig.java

```diff
@@ -1619,4 +1622,73 @@ public void checkSourceRootExistence() throws IOException {
             throw new IOException(String.format("Source root path \"%s\" is not readable", sourceRootPathFile.getAbsolutePath()));
         }
     }
+
+    /**
+     * Get all project related messages. These include
+     * <ol>
+     * <li>Main messages</li>
+     * <li>Messages with tag = project name</li>
+     * <li>Messages with tag = project's groups names</li>
+     * </ol>
+     *
+     * @return the sorted set of messages according to the accept time
+     * @see org.opengrok.indexer.web.messages.MessagesContainer#MESSAGES_MAIN_PAGE_TAG
+     */
+    private SortedSet<AcceptedMessage> getProjectMessages() {
+        SortedSet<AcceptedMessage> messages = getMessages();
+
+        if (getProject() != null) {
+            messages.addAll(getMessages(getProject().getName()));
+            getProject().getGroups().forEach(group -> {
+                messages.addAll(getMessages(group.getName()));
+            });
+        }
+
+        return messages;
+    }
+
+    /**
+     * Decide if this resource has been modified since the header value in the request.
+     * <p>
+     * The resource is modified since the weak etag value in the reuqest, the etag is
+     * computed using:
+     * <ul>
+     * <li>the source file modification</li>
+     * <li>project messages</li>
+     * <li>last timestamp for index</li>
+     * <li>OpenGrok current deployed version</li>
+     * </ul>
+     * <p>
+     * <p>
+     * If the resource was modified, appropriate headers in the response are filled.
+     *
+     * @param response the http response for setting the headers
+     * @return true if resource was not modified; false otherwise
+     */
+    public boolean isNotModified(HttpServletRequest request, HttpServletResponse response) {
+        String currentEtag = String.format("W/\"%s\"",
+                Objects.hash(
+                        // last modified time as UTC timestamp in millis
+                        getLastModified(),
+                        // all project related messages which changes the view
+                        getProjectMessages(),
+                        // last timestamp value
+                        getEnv().getDateForLastIndexRun().getTime(),
+                        // OpenGrok version has changed since the last time
+                        Info.getVersion()
+                )
+        );
+
+        String headerEtag = request.getHeader("If-None-Match");
```

Comment:
Is the literal value of the header name perhaps available in a pre-existing class ?

Reply:
yes